### PR TITLE
Implement flattening of row major matrix indexing

### PR DESCRIPTION
**Pull request by zeux**
To extract a column from row-major matrix, we need to do a strided load one
component at a time. In this case flattened_access_chain_offset still returns
the offset to the first element, but the stride is equal to matrix stride
instead of vector stride.

For this to work, we need to pass matrix stride (and transpose flag) through,
similar to how matrix flattening works.

Additionally slightly clean up recursive flattened_access_chain structure -
specifically, instead of deciding mid-traversal that we need matrix stride
information, we can just pass the matrix stride through - for access chains
that end in matrix/vector this gets us what we need, and for access chains
that end in structs the flattened_access_chain_struct code will recompute
correct stride/transposition data to pass through further.

### CODE REVIEW

File: reference/shaders/flatten/matrixindex.flatten.vert [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+out vec4 oE;
+
+void main()
+{
+    gl_Position = vec4(0.0);
+    oA = UBO[1];
+    oB = vec4(UBO[4].y, UBO[5].y, UBO[6].y, UBO[7].y);
+    oC = UBO[9];
+    oD = vec4(UBO[10].x, UBO[11].x, UBO[12].x, UBO[13].x);
+    oE = vec4(UBO[1].z, float(UBO[6].y), UBO[9].z, float(UBO[12].y));
```
Comment (HansKristian-ARM) [inline reference/shaders/flatten/matrixindex.flatten.vert]: Any idea why this dummy float(UBO[6].y) happens?

File: spirv_glsl.cpp [hunk 2]
```diff
@@ -3399,35 +3397,70 @@ std::string CompilerGLSL::flattened_access_chain_matrix(uint32_t base, const uin
...
+	if (need_transpose)
+	{
+		std::string expr;
 
-	expr += buffer_name;
-	expr += "[";
-	expr += result.first; // this is a series of N1 * k1 + N2 * k2 + ... that is either empty or ends with a +
-	expr += convert_to_string(index / 4);
-	expr += "]";
+		expr += type_to_glsl_constructor(target_type);
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Don't need constructor for scalar results.

File: reference/shaders/flatten/matrixindex.flatten.vert (hunk 1 above)
Comment (zeux) [inline reference/shaders/flatten/matrixindex.flatten.vert]: Ah, this is the constructor below I think. Will fix!

**Comment by HansKristian-ARM** [comment]
Linux passed, OSX failed due to a Travis error, so I'll just merge it.

**Comment by zeux** [comment]
Looks like it timed out. Travis OSX builds are much slower than Linux builds since they constantly have a backlog of builds :( We'll need to keep OSX builds if we want to test MSL thoroughly - I'm planning to submit a PR soon that validates MSL shaders using Metal compiler on OSX.

**Comment by HansKristian-ARM** [comment]
Cool. I'd like to see the MSL test suite more fleshed out, so that's nice.

**Pull request merged**